File: ports/jagentic-core/src/test/java/org/jagentic/core/WebInferenceTest.java
package org.jagentic.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

import com.sun.net.httpserver.HttpServer;

import org.junit.jupiter.api.Test;

import org.jagentic.core.inference.ClassifierGuardrail;
import org.jagentic.core.inference.ClassifierScorer;
import org.jagentic.core.inference.EmbeddingClassifier;
import org.jagentic.core.inference.LexiconClassifier;
import org.jagentic.core.web.WebFetcher;

/** Phase E — web toolkit (robots-aware fetch + link extraction against an in-process
 * HttpServer) and the DL inference SPI (lexicon + nearest-centroid classifiers, classifier
 * guardrail). All offline/deterministic. */
class WebInferenceTest {

  private static final String PAGE =
      "<html><head><title>Acme Bank</title><style>.x{color:red}</style></head>"
          + "<body><h1>Welcome</h1><p>Open a savings account today.</p>"
          + "<script>var x=1;</script>"
          + "<a href='/cards'>Cards</a> <a href='https://example.org/help'>Help</a>"
          + "</body></html>";

  private static HttpServer startServer(AtomicReference<String> robots) throws IOException {
    HttpServer server = HttpServer.create(new InetSocketAddress("127.0.0.1", 0), 0);
    server.createContext("/robots.txt", ex -> respond(ex, robots.get(), "text/plain"));
    server.createContext("/", ex -> respond(ex, PAGE, "text/html; charset=utf-8"));
    server.start();
    return server;
  }

  private static void respond(com.sun.net.httpserver.HttpExchange ex, String body, String ct) throws IOException {
    byte[] bytes = body.getBytes(StandardCharsets.UTF_8);
    ex.getResponseHeaders().add("Content-Type", ct);
    ex.sendResponseHeaders(200, bytes.length);
    try (OutputStream os = ex.getResponseBody()) {
      os.write(bytes);
    }
  }

  @Test
  void webFetchExtractsTextAndLinks() throws IOException {
    AtomicReference<String> robots = new AtomicReference<>("User-agent: *\nAllow: /\n");
    HttpServer server = startServer(robots);
    try {
      String base = "http://127.0.0.1:" + server.getAddress().getPort();
      WebFetcher.FetchResult r = new WebFetcher().fetch(base + "/page");
      assertEquals("Acme Bank", r.title());
      assertTrue(r.text().contains("Open a savings account today."), r.text());
      assertFalse(r.text().contains("var x"));
      assertFalse(r.text().contains("color:red"));
      assertTrue(r.links().contains(base + "/cards"), r.links().toString());
      assertTrue(r.links().contains("https://example.org/help"));
    } finally {
      server.stop(0);
    }
  }

  @Test
  void webToolsRegisterAndRun() throws IOException {
    AtomicReference<String> robots = new AtomicReference<>("User-agent: *\nAllow: /\n");
    HttpServer server = startServer(robots);
    try {
      String base = "http://127.0.0.1:" + server.getAddress().getPort();
      ToolRegistry reg = new ToolRegistry();
      List<String> ids = new WebFetcher().register(reg);
      assertTrue(ids.contains("web_fetch") && ids.contains("web_links"));
      String text = (String) reg.execute("web_fetch", Map.of("url", base + "/page"));
      assertTrue(text.contains("Acme Bank") && text.contains("savings account"));
      String links = (String) reg.execute("web_links", Map.of("url", base + "/page"));
      assertTrue(links.contains("https://example.org/help"));
    } finally {
      server.stop(0);
    }
  }

  @Test
  void robotsDisallowBlocksFetch() throws IOException {
    AtomicReference<String> robots = new AtomicReference<>("User-agent: *\nDisallow: /blocked\n");
    HttpServer server = startServer(robots);
    try {
      String base = "http://127.0.0.1:" + server.getAddress().getPort();
      WebFetcher f = new WebFetcher();
      assertThrows(SecurityException.class, () -> f.fetch(base + "/blocked"));
      assertEquals("Acme Bank", f.fetch(base + "/page").title());
    } finally {
      server.stop(0);
    }
  }

  @Test
  void lexiconClassifierLabelsByKeywords() {
    LexiconClassifier clf = new LexiconClassifier(Map.of(
        "billing", List.of("invoice", "charge", "refund", "payment"),
        "tech", List.of("error", "crash", "bug", "login")), "other");
    assertEquals("billing", clf.classify("I want a refund for this charge").label());
    assertEquals("tech", clf.classify("the app keeps showing a login error").label());
    assertEquals("other", clf.classify("hello there").label());
    double sum = clf.classify("refund my payment please").scores().values().stream().mapToDouble(d -> d).sum();
    assertTrue(Math.abs(sum - 1.0) < 1e-9, "scores sum=" + sum);
  }

  @Test
  void embeddingClassifierNearestCentroid() {
    EmbeddingClassifier clf = new EmbeddingClassifier().fit(Map.of(
        "billing", List.of("refund my invoice", "dispute a charge", "payment failed"),
        "tech", List.of("app crashed on login", "error message bug", "cannot sign in")));
    assertEquals("billing", clf.classify("please refund this invoice charge").label());
    assertEquals("tech", clf.classify("login error keeps crashing").label());
    double sum = clf.classify("refund invoice").scores().values().stream().mapToDouble(d -> d).sum();
    assertTrue(Math.abs(sum - 1.0) < 1e-6, "scores sum=" + sum);
  }

  @Test
  void classifierGuardrailBlocksLabel() {
    LexiconClassifier clf = new LexiconClassifier(Map.of(
        "toxic", List.of("idiot", "stupid", "hate"),
        "ok", List.of("please", "thanks", "help")), "ok");
    ClassifierGuardrail guard = new ClassifierGuardrail(clf, List.of("toxic"), 0.3, null, false);
    assertNotNull(guard.checkInput("you stupid idiot"));
    assertNull(guard.checkInput("please help, thanks"));
    ClassifierScorer scorer = new ClassifierScorer(clf, "toxic");
    assertTrue(scorer.score("idiot") > scorer.score("thanks please help"));
  }
}

File: ports/experimental/go/core/web_inference_test.go
package core

import (
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

const webPage = `<html><head><title>Acme Bank</title><style>.x{color:red}</style></head>` +
	`<body><h1>Welcome</h1><p>Open a savings account today.</p>` +
	`<script>var x=1;</script>` +
	`<a href="/cards">Cards</a> <a href="https://example.org/help">Help</a></body></html>`

func webServer(robots string) *httptest.Server {
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/robots.txt":
			w.Header().Set("Content-Type", "text/plain")
			_, _ = w.Write([]byte(robots))
		default:
			w.Header().Set("Content-Type", "text/html; charset=utf-8")
			_, _ = w.Write([]byte(webPage))
		}
	}))
}

func TestWebFetchExtractsTextAndLinks(t *testing.T) {
	srv := webServer("User-agent: *\nAllow: /\n")
	defer srv.Close()

	f := NewWebFetcher()
	res, err := f.Fetch(srv.URL + "/page")
	if err != nil {
		t.Fatalf("fetch: %v", err)
	}
	if res.Title != "Acme Bank" {
		t.Fatalf("title = %q", res.Title)
	}
	if !strings.Contains(res.Text, "Open a savings account today.") {
		t.Fatalf("text missing body: %q", res.Text)
	}
	if strings.Contains(res.Text, "var x") || strings.Contains(res.Text, "color:red") {
		t.Fatalf("script/style leaked into text: %q", res.Text)
	}
	foundRel, foundAbs := false, false
	for _, l := range res.Links {
		if l == srv.URL+"/cards" {
			foundRel = true
		}
		if l == "https://example.org/help" {
			foundAbs = true
		}
	}
	if !foundRel || !foundAbs {
		t.Fatalf("links = %v", res.Links)
	}
}

func TestWebToolsRegisterAndRun(t *testing.T) {
	srv := webServer("User-agent: *\nAllow: /\n")
	defer srv.Close()

	reg := NewToolRegistry()
	ids := RegisterWebTools(reg)
	if len(ids) != 2 || !reg.Has("web_fetch") || !reg.Has("web_links") {
		t.Fatalf("register ids = %v", ids)
	}
	text := reg.Execute("web_fetch", map[string]any{"url": srv.URL + "/page"}).(string)
	if !strings.Contains(text, "Acme Bank") || !strings.Contains(text, "savings account") {
		t.Fatalf("web_fetch = %q", text)
	}
	links := reg.Execute("web_links", map[string]any{"url": srv.URL + "/page"}).(string)
	if !strings.Contains(links, "https://example.org/help") {
		t.Fatalf("web_links = %q", links)
	}
}

func TestRobotsDisallowBlocksFetch(t *testing.T) {
	srv := webServer("User-agent: *\nDisallow: /blocked\n")
	defer srv.Close()

	f := NewWebFetcher()
	if _, err := f.Fetch(srv.URL + "/blocked"); err == nil {
		t.Fatal("expected robots disallow error")
	}
	if _, err := f.Fetch(srv.URL + "/page"); err != nil {
		t.Fatalf("allowed path should fetch: %v", err)
	}
}

func TestLexiconClassifier(t *testing.T) {
	clf := NewLexiconClassifier(map[string][]string{
		"billing": {"invoice", "charge", "refund", "payment"},
		"tech":    {"error", "crash", "bug", "login"},
	}, "other")
	if got := clf.Classify("I want a refund for this charge").Label; got != "billing" {
		t.Fatalf("billing => %q", got)
	}
	if got := clf.Classify("the app keeps showing a login error").Label; got != "tech" {
		t.Fatalf("tech => %q", got)
	}
	if got := clf.Classify("hello there").Label; got != "other" {
		t.Fatalf("other => %q", got)
	}
	c := clf.Classify("refund my payment please")
	sum := 0.0
	for _, v := range c.Scores {
		sum += v
	}
	if sum < 0.999 || sum > 1.001 {
		t.Fatalf("scores not a distribution: %v (sum %f)", c.Scores, sum)
	}
}

func TestEmbeddingClassifierNearestCentroid(t *testing.T) {
	clf, err := NewEmbeddingClassifier(nil, 0).Fit(map[string][]string{
		"billing": {"refund my invoice", "dispute a charge", "payment failed"},
		"tech":    {"app crashed on login", "error message bug", "cannot sign in"},
	})
	if err != nil {
		t.Fatalf("fit: %v", err)
	}
	if got := clf.Classify("please refund this invoice charge").Label; got != "billing" {
		t.Fatalf("billing => %q", got)
	}
	if got := clf.Classify("login error keeps crashing").Label; got != "tech" {
		t.Fatalf("tech => %q", got)
	}
}

func TestClassifierGuardrailBlocksLabel(t *testing.T) {
	clf := NewLexiconClassifier(map[string][]string{
		"toxic": {"idiot", "stupid", "hate"},
		"ok":    {"please", "thanks", "help"},
	}, "ok")
	guard := NewClassifierGuardrail(clf, []string{"toxic"}, 0.3, "", false)
	if guard.CheckInput("you stupid idiot") == "" {
		t.Fatal("expected toxic block")
	}
	if guard.CheckInput("please help, thanks") != "" {
		t.Fatal("benign text should pass")
	}
	scorer := NewClassifierScorer(clf, "toxic")
	if scorer.Score("idiot") <= scorer.Score("thanks please help") {
		t.Fatal("toxic score should exceed benign")
	}
}
